### DISCUSSION TITLE
openssl 3.0 support

### DISCUSSION THREAD

**Discussion by rikeshshakyasz** [Azure/azure-iot-sdk-c#2596] [General]
I use openssl 3.0 and I keep getting errors for the use of openssl 3.0 deprecated functions. Is openssl 3.0 supported by the new LTS_03_2024 release?

**Comment by ericwolz** [comment]
openssl 3.0 is supported through the 1.1.1 interface that it provides.

**Reply by rikeshshakyasz** [reply]
would it be possible for you to elaborate on  how to achieve this?

**Comment by ericwolz** [comment]
openssl 3.0 library is already supported

https://github.com/Azure/azure-c-shared-utility/discussions/566

**Reply by rikeshshakyasz** [reply]
I tried using #define OPENSSL_SUPPRESS_DEPRECATED 1 on files x509_openssl.c and tlsio_openssl.c. However, I still got errors for if (ssl_ctx->extra_certs != NULL) in x509_openssl.c. Do I need to add any compilation flags to get this working?

**Comment by ericwolz** [comment]
Maybe if your share the actual errors people can help you better.

**Reply by rikeshshakyasz** [reply]
Here is the actual error I was getting,

azure-iot-sdk-c-2024-03-04/c-utility/adapters/tlsio_openssl.c: In function 'engine_destroy':
azure-iot-sdk-c-2024-03-04/c-utility/adapters/tlsio_openssl.c:762:9: error: 'ENGINE_free' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  762 |         ENGINE_free(tls->engine); // Release structural reference.
      |         ^~~~~~~~~~~
In file included from azure-iot-sdk-c-2024-03-04/c-utility/adapters/tlsio_openssl.c:9:
/target-mips_24kc_musl/usr/include/openssl/engine.h:493:27: note: declared here
  493 | OSSL_DEPRECATEDIN_3_0 int ENGINE_free(ENGINE *e);
      |                           ^~~~~~~~~~~
azure-iot-sdk-c-2024-03-04/c-utility/adapters/tlsio_openssl.c: In function 'engine_load':
azure-iot-sdk-c-2024-03-04/c-utility/adapters/tlsio_openssl.c:772:5: error: 'ENGINE_by_id' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  772 |     tls->engine = ENGINE_by_id(tls->engine_id);
      |     ^~~
/target-mips_24kc_musl/usr/include/openssl/engine.h:336:31: note: declared here
  336 | OSSL_DEPRECATEDIN_3_0 ENGINE *ENGINE_by_id(const char *id);
      |                               ^~~~~~~~~~~~
azure-iot-sdk-c-2024-03-04/c-utility/adapters/tlsio_openssl.c: In function 'add_certificate_to_store':
azure-iot-sdk-c-2024-03-04/c-utility/adapters/tlsio_openssl.c:961:24: error: assignment discards 'const' qualifier from pointer target type [-Werror=discarded-qualifiers]
  961 |             bio_method = BIO_s_mem();
      |                        ^
azure-iot-sdk-c-2024-03-04/c-utility/adapters/tlsio_openssl.c: In function 'create_openssl_instance':
azure-iot-sdk-c-2024-03-04/c-utility/adapters/tlsio_openssl.c:1058:9: error: 'TLSv1_2_method' is deprecated: Since OpenSSL 1.1.0 [-Werror=deprecated-declarations]
 1058 |         method = TLSv1_2_method();
      |         ^~~~~~
In file included from azure-iot-sdk-c-2024-03-04/c-utility/adapters/tlsio_openssl.c:5:
/target-mips_24kc_musl/usr/include/openssl/ssl.h:2037:50: note: declared here
 2037 | OSSL_DEPRECATEDIN_1_1_0 __owur const SSL_METHOD *TLSv1_2_method(void); /* TLSv1.2 */
      |                                                  ^~~~~~~~~~~~~~
azure-iot-sdk-c-2024-03-04/c-utility/adapters/tlsio_openssl.c:1062:9: error: 'TLSv1_1_method' is deprecated: Since OpenSSL 1.1.0 [-Werror=deprecated-declarations]
 1062 |         method = TLSv1_1_method();
      |         ^~~~~~
/target-mips_24kc_musl/usr/include/openssl/ssl.h:2029:50: note: declared here
 2029 | OSSL_DEPRECATEDIN_1_1_0 __owur const SSL_METHOD *TLSv1_1_method(void); /* TLSv1.1 */
      |                                                  ^~~~~~~~~~~~~~
azure-iot-sdk-c-2024-03-04/c-utility/adapters/tlsio_openssl.c:1066:9: error: 'TLSv1_method' is deprecated: Since OpenSSL 1.1.0 [-Werror=deprecated-declarations]
 1066 |         method = TLSv1_method();
      |         ^~~~~~
/target-mips_24kc_musl/usr/include/openssl/ssl.h:2021:50: note: declared here
 2021 | OSSL_DEPRECATEDIN_1_1_0 __owur const SSL_METHOD *TLSv1_method(void); /* TLSv1.0 */
      |                                                  ^~~~~~~~~~~~
azure-iot-sdk-c-2024-03-04/c-utility/adapters/tlsio_openssl.c: In function 'tlsio_openssl_init':
azure-iot-sdk-c-2024-03-04/c-utility/adapters/tlsio_openssl.c:1224:5: error: 'ERR_load_BIO_strings' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
 1224 |     ERR_load_BIO_strings();
      |     ^~~~~~~~~~~~~~~~~~~~
In file included from /target-mips_24kc_musl/usr/include/openssl/cryptoerr.h:17,
                 from /target-mips_24kc_musl/usr/include/openssl/crypto.h:38,
                 from /target-mips_24kc_musl/usr/include/openssl/bio.h:30,
                 from /target-mips_24kc_musl/usr/include/openssl/ssl.h:29:
/target-mips_24kc_musl/usr/include/openssl/cryptoerr_legacy.h:31:27: note: declared here
   31 | OSSL_DEPRECATEDIN_3_0 int ERR_load_BIO_strings(void);
      |                           ^~~~~~~~~~~~~~~~~~~~
azure-iot-sdk-c-2024-03-04/c-utility/adapters/tlsio_openssl.c: In function 'tlsio_openssl_deinit':
azure-iot-sdk-c-2024-03-04/c-utility/adapters/tlsio_openssl.c:1252:5: error: 'ERR_remove_thread_state' is deprecated: Since OpenSSL 1.1.0 [-Werror=deprecated-declarations]
 1252 |     ERR_remove_thread_state(NULL);
      |     ^~~~~~~~~~~~~~~~~~~~~~~
In file included from azure-iot-sdk-c-2024-03-04/c-utility/adapters/tlsio_openssl.c:6:
/target-mips_24kc_musl/usr/include/openssl/err.h:485:30: note: declared here
  485 | OSSL_DEPRECATEDIN_1_1_0 void ERR_remove_thread_state(void *);
      |                              ^~~~~~~~~~~~~~~~~~~~~~~
azure-iot-sdk-c-2024-03-04/c-utility/adapters/tlsio_openssl.c: In function 'tlsio_openssl_setoption':
azure-iot-sdk-c-2024-03-04/c-utility/adapters/tlsio_openssl.c:1724:13: error: 'ENGINE_load_builtin_engines' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
 1724 |             ENGINE_load_builtin_engines();
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/target-mips_24kc_musl/usr/include/openssl/engine.h:358:28: note: declared here
  358 | OSSL_DEPRECATEDIN_3_0 void ENGINE_load_builtin_engines(void);
  
  
  azure-iot-sdk-c-2024-03-04/c-utility/adapters/x509_openssl.c: In function 'load_certificate_chain':
azure-iot-sdk-c-2024-03-04/c-utility/adapters/x509_openssl.c:81:28: error: invalid use of incomplete typedef 'SSL_CTX' {aka 'struct ssl_ctx_st'}
   81 |                 if (ssl_ctx->extra_certs != NULL)
      |                            ^~
In file included from /target-mips_24kc_musl/usr/include/openssl/ssl.h:31,
                 from azure-iot-sdk-c-2024-03-04/c-utility/inc/azure_c_shared_utility/x509_openssl.h:7,
                 from azure-iot-sdk-c-2024-03-04/c-utility/adapters/x509_openssl.c:4:
azure-iot-sdk-c-2024-03-04/c-utility/adapters/x509_openssl.c:83:45: error: invalid use of incomplete typedef 'SSL_CTX' {aka 'struct ssl_ctx_st'}
   83 |                     sk_X509_pop_free(ssl_ctx->extra_certs, X509_free);
      |                                             ^~
azure-iot-sdk-c-2024-03-04/c-utility/adapters/x509_openssl.c:84:28: error: invalid use of incomplete typedef 'SSL_CTX' {aka 'struct ssl_ctx_st'}
   84 |                     ssl_ctx->extra_certs = NULL;
      |                            ^~
azure-iot-sdk-c-2024-03-04/c-utility/adapters/x509_openssl.c: In function 'load_key_RSA':
azure-iot-sdk-c-2024-03-04/c-utility/adapters/x509_openssl.c:140:5: error: 'EVP_PKEY_get1_RSA' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  140 |     RSA* privatekey = EVP_PKEY_get1_RSA(evp_key);
      |     ^~~
In file included from /target-mips_24kc_musl/usr/include/openssl/x509.h:29:
/target-mips_24kc_musl/usr/include/openssl/evp.h:1350:16: note: declared here
 1350 | struct rsa_st *EVP_PKEY_get1_RSA(EVP_PKEY *pkey);
      |                ^~~~~~~~~~~~~~~~~
azure-iot-sdk-c-2024-03-04/c-utility/adapters/x509_openssl.c:150:9: error: 'SSL_CTX_use_RSAPrivateKey' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  150 |         if (SSL_CTX_use_RSAPrivateKey(ssl_ctx, privatekey) != 1)
      |         ^~
/target-mips_24kc_musl/usr/include/openssl/ssl.h:1816:12: note: declared here
 1816 | __owur int SSL_CTX_use_RSAPrivateKey(SSL_CTX *ctx, RSA *rsa);
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~
azure-iot-sdk-c-2024-03-04/c-utility/adapters/x509_openssl.c:162:9: error: 'RSA_free' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  162 |         RSA_free(privatekey);
      |         ^~~~~~~~
In file included from /target-mips_24kc_musl/usr/include/openssl/x509.h:36:
/target-mips_24kc_musl/usr/include/openssl/rsa.h:293:28: note: declared here
  293 | OSSL_DEPRECATEDIN_3_0 void RSA_free(RSA *r);
      |                            ^~~~~~~~
azure-iot-sdk-c-2024-03-04/c-utility/adapters/x509_openssl.c: In function 'x509_openssl_add_engine_key':
azure-iot-sdk-c-2024-03-04/c-utility/adapters/x509_openssl.c:227:5: error: 'ENGINE_init' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  227 |     if (!ENGINE_init(engine))
      |     ^~
In file included from azure-iot-sdk-c-2024-03-04/c-utility/adapters/x509_openssl.c:14:
/target-mips_24kc_musl/usr/include/openssl/engine.h:620:27: note: declared here
  620 | OSSL_DEPRECATEDIN_3_0 int ENGINE_init(ENGINE *e);
      |                           ^~~~~~~~~~~
azure-iot-sdk-c-2024-03-04/c-utility/adapters/x509_openssl.c:234:9: error: 'ENGINE_set_default' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  234 |         if (!ENGINE_set_default(engine, ENGINE_METHOD_ALL))
      |         ^~
/target-mips_24kc_musl/usr/include/openssl/engine.h:708:27: note: declared here
  708 | OSSL_DEPRECATEDIN_3_0 int ENGINE_set_default(ENGINE *e, unsigned int flags);
      |                           ^~~~~~~~~~~~~~~~~~
azure-iot-sdk-c-2024-03-04/c-utility/adapters/x509_openssl.c:241:13: error: 'ENGINE_load_private_key' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  241 |             EVP_PKEY* evp_key = ENGINE_load_private_key(engine, x509privatekey_id, NULL, NULL);
      |             ^~~~~~~~
/target-mips_24kc_musl/usr/include/openssl/engine.h:638:11: note: declared here
  638 | EVP_PKEY *ENGINE_load_private_key(ENGINE *e, const char *key_id,
      |           ^~~~~~~~~~~~~~~~~~~~~~~
azure-iot-sdk-c-2024-03-04/c-utility/adapters/x509_openssl.c:264:9: error: 'ENGINE_finish' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  264 |         if (!ENGINE_finish(engine))  // Release functional reference.
      |         ^~
/target-mips_24kc_musl/usr/include/openssl/engine.h:628:27: note: declared here
  628 | OSSL_DEPRECATEDIN_3_0 int ENGINE_finish(ENGINE *e);
      |                           ^~~~~~~~~~~~~
azure-iot-sdk-c-2024-03-04/c-utility/adapters/x509_openssl.c: In function 'x509_openssl_add_certificates':
azure-iot-sdk-c-2024-03-04/c-utility/adapters/x509_openssl.c:353:24: error: assignment discards 'const' qualifier from pointer target type [-Werror=discarded-qualifiers]
  353 |             bio_method = BIO_s_mem();

**Comment by ericwolz** [comment]
So these are the compile errors I see above. Looks like there is a compatibility issue with your mips compiler. Please look at the following thread to see if that helps.

https://github.com/monero-project/monero/issues/2799


```
azure-iot-sdk-c-2024-03-04/c-utility/adapters/tlsio_openssl.c: In function 'add_certificate_to_store':
azure-iot-sdk-c-2024-03-04/c-utility/adapters/tlsio_openssl.c:961:24: error: assignment discards 'const' qualifier from pointer target type [-Werror=discarded-qualifiers]
961 | bio_method = BIO_s_mem();


azure-iot-sdk-c-2024-03-04/c-utility/adapters/x509_openssl.c: In function 'load_certificate_chain':
azure-iot-sdk-c-2024-03-04/c-utility/adapters/x509_openssl.c:81:28: error: invalid use of incomplete typedef 'SSL_CTX' {aka 'struct ssl_ctx_st'}
81 | if (ssl_ctx->extra_certs != NULL)
| ^~
In file included from /target-mips_24kc_musl/usr/include/openssl/ssl.h:31,
from azure-iot-sdk-c-2024-03-04/c-utility/inc/azure_c_shared_utility/x509_openssl.h:7,
from azure-iot-sdk-c-2024-03-04/c-utility/adapters/x509_openssl.c:4:
azure-iot-sdk-c-2024-03-04/c-utility/adapters/x509_openssl.c:83:45: error: invalid use of incomplete typedef 'SSL_CTX' {aka 'struct ssl_ctx_st'}
83 | sk_X509_pop_free(ssl_ctx->extra_certs, X509_free);
| ^~
azure-iot-sdk-c-2024-03-04/c-utility/adapters/x509_openssl.c:84:28: error: invalid use of incomplete typedef 'SSL_CTX' {aka 'struct ssl_ctx_st'}
84 | ssl_ctx->extra_certs = NULL;
| ^~


azure-iot-sdk-c-2024-03-04/c-utility/adapters/x509_openssl.c: In function 'x509_openssl_add_certificates':
azure-iot-sdk-c-2024-03-04/c-utility/adapters/x509_openssl.c:353:24: error: assignment discards 'const' qualifier from pointer target type [-Werror=discarded-qualifiers]
353 | bio_method = BIO_s_mem();

```

**Reply by rikeshshakyasz** [reply]
Unfortunately, that does not seem to help, it seems to me that I am not setting some configurations. However, I am unsure on what that is as I am a newbie on this.

**Comment by ericwolz** [comment]
This is our MIPS cross compile and this works. Please look at it to see if that helps.

https://github.com/Azure/azure-iot-sdk-c/blob/main/samples/dockerbuilds/MIPS32/Dockerfile
